_posts/2016-08-22-the-java-synchronisers.md

```diff
@@ -0,0 +1,111 @@
+
+### Semaphore
+
+The [Semaphore](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html) is used to restrict threads access to a certain resource. It is initialised with a number of permits. 
+
+The `acquire` method **blocks** until a permit is available and takes it.
+The `release` method **adds** a permit, releasing a blocking acquierer. Note that calls to release does not have to be made by the same thread that called acquire.
+
+The semaphore can be **fair** or **unfair**. If fair, then the threads acquire permits in a FIFO fashion.
+
+Although at first it may seem similar to the CountDownLatch its purpose is completely different.
+
+### CyclicBarrier
+
+The [CyclicBarrier](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CyclicBarrier.html) is built around the concept of **parties**. It allows threads to wait for each other to reach a common barrier point.
+
+The `await` method **blocks** until all parties arrive. It behaves somehow as the inverse of the *CountDownLatch*. After *N* awaits it continues.
+
+It has support for an optiona runnable that runs once per barrier point. After last party arrives, but before they are released. It is usually used to **update** shared-state between threads.
+
+It is cyclir because it can be *reused* after threads are released.
+
+### Exchanger
+
+The [Exchanger](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Exchanger.html) is a synchronization point at which two threads can exchange information.
+
+Threads **block** until its counterpart presents its information. The same behaviour occurs at both sides
+
+### Phaser
+
+The [Phaser](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Phaser.html) is a reusable barrier, similar to *CountDownLatch* and *CyclirBarrier*, but much more flexible.
+
+In the phaser the number of registered parties is not fixed at creation time. Parties can register at any time through `register` or `bulkRegister` methods. Parties can deregister upon arrival with `arriveAndDeregister`.
+
+If offers several methods for synchronization. The `arriveAndAwaitAdvance` method behaves the same way as *CycleBarrier* `await` method does. `arrive` and `arriveAndDeregister` record arrival, but don't block. `awaitAdvance` blocks until all parties arrive.
+
+It can be terminated, forcing all synchronization methods to return. Can be forced through the `forceTermination` method.
+
+It also provides support for monitoring its state. It is noteworthy mentioning that synchronization methods can be called only by registered parties, whilst state can be monitored by any caller. Monitoring methods include `getRegisteredParties` and `getArrivedParties` among others.
+
+## Conclusion
+
+Multithreading is definitely not an easy problem, but can be somehow easier to tackle using the tools that languages provide.
```

Comment:
that languages provide. => some of the languages provide.

(In my view not every language provide supporting tools).